src/plugins/formulas/formulas.js

```diff
@@ -88,356 +38,297 @@ export class Formulas extends BasePlugin {
+    };
 
-    if (isFormulaExpressionEscaped(renderValue)) {
-      renderValue = unescapeFormulaExpression(renderValue);
-    }
+    if (ioMode === 'get') {
+      const cellValue = this.hyperformula.getCellValue(address);
 
-    return renderValue;
-  }
+      // If `cellValue` is an object it is expected to be an error
+      const value = (typeof cellValue === 'object' && cellValue !== null) ? cellValue.value : cellValue;
 
-  /**
-   * On before validate listener.
-   *
-   * @private
-   * @param {*} value Value to validate.
-   * @param {number} row Row index.
-   * @param {number} prop Column property.
-   * @returns {*}
-   */
-  onBeforeValidate(value, row, prop) {
-    const column = this.hot.propToCol(prop);
-    let validateValue = value;
+      // Omit the leading `'` from presentation, and all `getData` operations
+      const prettyValue = (() => {
+        if (typeof value === 'string') {
+          return value.indexOf('\'') === 0 ? value.slice(1) : value;
+        }
 
-    if (this.hasComputedCellValue(row, column)) {
-      validateValue = this.getCellValue(row, column);
-    }
+        return value;
+      })();
 
-    return validateValue;
+      valueHolder.value = prettyValue;
+    } else {
+      this.hyperformula.setCellContents(address, valueHolder.value);
+    }
   }
 
-  /**
-   * `afterSetDataAtCell` listener.
-   *
-   * @private
-   * @param {Array} changes Array of changes.
-   * @param {string} [source] Source of changes.
-   */
-  onAfterSetDataAtCell(changes, source) {
-    if (source === 'loadData') {
+  onModifySourceData(row, col, valueHolder, ioMode) {
+    if (!this.enabled) {
       return;
     }
 
-    this.dataProvider.clearChanges();
-
-    arrayEach(changes, ([row, column, oldValue, newValue]) => {
-      const physicalColumn = this.hot.propToCol(column);
-      const physicalRow = this.hot.toPhysicalRow(row);
-      let value = newValue;
+    const dimensions = this.hyperformula.getSheetDimensions(this.hyperformula.getSheetId(this.sheetName));
 
-      if (isFormulaExpression(value)) {
-        value = toUpperCaseFormula(value);
-      }
+    // Don't actually change the source data if HyperFormula is not
+    // initialized yet. This is done to allow the `afterLoadData` hook to
+    // load the existing source data with `Handsontable#getSourceDataArray`
+    // properly.
+    if (dimensions.width === 0 && dimensions.height === 0) {
+      return;
+    }
 
-      this.dataProvider.collectChanges(physicalRow, physicalColumn, value);
+    const address = {
+      row: this.hot.toVisualRow(row),
+      col,
+      sheet: this.hyperformula.getSheetId(this.sheetName)
+    };
 
-      if (oldValue !== value) {
-        this.sheet.applyChanges(physicalRow, physicalColumn, value);
-      }
-    });
-    this.recalculate();
+    if (ioMode === 'get') {
+      valueHolder.value = this.hyperformula.getCellSerialized(address);
+    } else if (ioMode === 'set') {
```

Comment:
A new line before `else if` would make it more readable.
```suggestion

    } else if (ioMode === 'set') {
```

Reply:
Maybe we should update from the develop? We have ESLint rule and an autofix for this now

Reply:
No eslint errors reported :shrug: https://github.com/handsontable/handsontable/runs/2450017513